Feature/native wrapper pr

nonec opened the pull request:
I wanted to try the native wrapper for a project. So I updated dependencies and made adjustments to make it work for me. It's also using Swift 5 now.
In the end I tried to clean it up but there might be some more necessary.
Made it run on Travis CI and adjusted the tests to pass.

I would be happy to receive feedback :-)

stephengroat committed:
Create Brewfile

stephengroat committed:
move to brew bundle

stephengroat committed:
move to brew update

Commit:
Merge pull request #34 from stephengroat/patch-1

Improves travisCI steps

Commit:
wait on verify to complete before allowing tests to continue to prevent race condition

Commit:
Update BrightFutures

andrewspinks committed:
Merge pull request #37 from rvoortman/update-bright-futures

Update BrightFutures

andrewspinks committed:
Merge pull request #38 from stitchfix/master

Wait on verify to complete before allowing tests to continue

andrewspinks committed:
Update gems

andrewspinks committed:
Update release script to work better with cocoa pods

andrewspinks committed:
update changelog for release 0.5.2

andrewspinks committed:
bump cocoapods version to 0.5.2

surpher committed:
Updates README to encourage usage of pact-ruby-standalone

surpher committed:
Removed reference to non-standalone pact mock service

andrewspinks committed:
Merge pull request #40 from surpher/master

Updates README to encourage usage of pact-ruby-standalone

surpher committed:
Update readme

surpher committed:
Update BrightFutures dependency version

surpher committed:
Project updates for Swift 4.2

surpher committed:
Updates travis pipeline to run tests on iOS 12.1 and 11.3

surpher committed:
Bump up Alamofire version

surpher committed:
Bumps up versions of all dependencies

surpher committed:
Update travis CI env matrix

surpher committed:
Update Quick dependency to use “master”

(https://github.com/Quick/Nimble/issues/608)

surpher committed:
Update Cartfile.resolved for use with bootstrap on CI

andrewspinks committed:
Merge pull request #42 from surpher/master

Update project and dependencies for Swift 4.2

andrewspinks committed:
Update pact-mock-service versions

andrewspinks committed:
Update cocoapods dependencies.

andrewspinks committed:
update changelog for release 0.5.3

andrewspinks committed:
bump cocoapods version to 0.5.3

andrewspinks committed:
Fix cocoa pods release version number and release script.

georgescumihai committed:
Update dependencies

andrewspinks committed:
Merge pull request #44 from georgescumihai/feature/update-dependencies

Update dependencies

Commit:
Merge branch 'master' into feature/nativeWrapper

# Conflicts:
#	.travis.yml
#	Cartfile
#	Cartfile.resolved
#	Gemfile
#	Gemfile.lock
#	Package.resolved
#	Package.swift
#	PactConsumerSwift.podspec
#	PactConsumerSwift.xcodeproj/project.pbxproj
#	PactConsumerSwift.xcodeproj/xcshareddata/xcschemes/PactConsumerSwift iOS.xcscheme
#	PactConsumerSwift.xcodeproj/xcshareddata/xcschemes/PactConsumerSwift macOS.xcscheme
#	PactConsumerSwift.xcodeproj/xcshareddata/xcschemes/PactConsumerSwift tvOS.xcscheme
#	Sources/Interaction.swift
#	Sources/MockService.swift
#	Sources/Ruby/RubyMockServer.swift

Commit:
synced Cartfile, SPM, podspec dependencies and versions

Commit:
resolved dependencies to work

Commit:
fixed swift linter errors

Commit:
removed result framework references from project file

Commit:
updated swiftlint version

Commit:
updated project to swift 5.0

Commit:
updated libpact_mock_server.a

Commit:
* made NativeMockServerWrapper usable

Commit:
* using the native matcher instead of the ruby matcher

Commit:
making sure the pact_mock_server header and binary are with the pod

Commit:
* reverted matcher override

Commit:
* simple solution to set the matcher

Commit:
updated libpact_mock_server.a (was built with: cargo lipo --release --targets armv7-apple-ios,x86_64-apple-ios)

Commit:
updated libpact_mock_server.a to inclued arm64 as well (current targets: armv7-apple-ios,x86_64-apple-ios,aarch64-apple-ios,armv7s-apple-ios)

Commit:
enabled user to set custom port to prevent collision for multiple instances

Commit:
* added handling for errors on write

Commit:
fixed crashes by returning on occuring errors

Commit:
gracefully handling non-existent directories

Commit:
* removed port setting ability from initializer
* added retries if port is blocked
* increased port range

Commit:
added some debug output

Commit:
show info on cleanup

Commit:
* added notify debug message if folder not found

Commit:
added logic to create pact path

Commit:
cleaned writing errors a bit

Commit:
introduced possibility to append interactions to pact

Commit:
introduced flag to decide if pact should be written

Commit:
minor changes to pact dir setting and error message text

Commit:
* fixed not waiting for verification while running pact test

Commit:
removed FIXMEs

Commit:
removed Result framework from Carthage scripts in test targets

Commit:
adjusted iOS version in build script

Commit:
fixed warning

Commit:
fixed warning

Commit:
added constructor with default matcher (to not break tests)

Commit:
removed extended file writing logic

Commit:
added order for dictionary keys (for the tests)

Commit:
added MinTypeMatcher

Commit:
added and adjusted to MinTypeMatcher changes

Commit:
using MinTypeMatcher in PactBodyBuilder

Commit:
add body of MinTypeMatcher multiple times

Commit:
removed warning

Commit:
adapted and added test for minType rule

Commit:
updated fastlane

Commit:
updated query test to pass

Commit:
updated CONTRIBUTING.md

Commit:
moved swift version to podspec

Commit:
added individual libraries for mac os and ios

Commit:
synced ios versions for default tests

Commit:
fixed target selection

Commit:
split up libs more

Commit:
moved lib files and added combine and copy scripts

Commit:
added SwiftyJSON as dependency to SPM

Commit:
point to current directory to satisfy pod spec lint

Commit:
trigger commit

Commit:
set swift version to 4.2 because 5.0 failed on travis

Commit:
faster builds first

Commit:
adjusted all targets to swift version 4.2

Commit:
Revert "trigger commit"

This reverts commit cdfc34d87e5a7010dcafd6208160082e2227dd37.

Commit:
increased xcode version and added swift version download

Commit:
readded swift 5 versions

Commit:
removed install step

Commit:
added bundle update

Commit:
recommended way of updating bundler

Commit:
readded xcpretty to Gemfile

Commit:
adjusted header search path in podspec

Commit:
removed duplicate header

andrewspinks commented:
Oh, wow, thanks @nonec . When I was playing with this ages ago there were a few issues which was blocking it from being merged into master (bitcode issues with rust, and I didn't get around to supporting the pact specification v2 matches). Could you change this pull request to go back into the original branch, and not master? Thanks!

nonec commented:
> Oh, wow, thanks @nonec . When I was playing with this ages ago there were a few issues which was blocking it from being merged into master (bitcode issues with rust, and I didn't get around to supporting the pact specification v2 matches). Could you change this pull request to go back into the original branch, and not master? Thanks!

Done!

nonec commented:
> Oh, wow, thanks @nonec . When I was playing with this ages ago there were a few issues which was blocking it from being merged into master (bitcode issues with rust, and I didn't get around to supporting the pact specification v2 matches). Could you change this pull request to go back into the original branch, and not master? Thanks!

I had to recompile the rust library (and split it to fit into git) and it seems to work (or I'm missing something obvious). Hope it'll be useful to the project.

nonec commented:
Also I changed the podspec for the lint and added a tag already. Not sure if that was the right way though. Get back to me if there's more feedback :-)

andrewspinks commented on Tests/Swift/e2e/PactConsumerSwiftTests/PactSpecs.swift:
```diff
@@ -131,7 +131,7 @@ class PactSwiftSpec: QuickSpec {
 
             it("should return animals living in water using matcher") {
-              let queryMatcher = testSetup.matcher.term(matcher: "live=*", generate: "live=water")
+                let queryMatcher = ["live": testSetup.matcher.term(matcher: ".*", generate: "water")]
```
Is there a reason for changing this test? We should maintain support for existing implementations, and this used to work (although it is a valid case, so maybe it should be a new test).

nonec replied:
~I remember that I had some problems with this test and I must have forgotten to revert it. I'll update this branch later and I don't have the problems there anymore.~

edit: I just looked into this. turns out the generation differs when using a String or a Dictionary for a query. With a String it results in `$query` key in the matching rule while as a Dictionary it results in `$query.live[0]` key. The zero is hardcoded for this in the `PactQueryBuilder`.
This is also made sure of by the existing `PactQueryBuilder` tests. 

It seems to me that the rust library isn't capable of working with the first approach for some reason.
`GET /animals: {error: Expected 'water' to match 'live=*', expected: water, actual: water}
`

This was the reason I changed it last year. I just didn't dive that deep into it at that point.
@andrewspinks can you clear up for me which behaviour is correct or if they both are my conclusions from this might be correct?

andrewspinks commented on Tests/Swift/e2e/PactConsumerSwiftTests/PactSpecs.swift:
```diff
@@ -141,7 +141,7 @@ class PactSwiftSpec: QuickSpec {
 
               //Run the tests
-              animalMockService!.run { (testComplete) -> Void in
+              animalMockService!.run(timeout: 5) { (testComplete) -> Void in
```
Any reason to add a timeout to this test?

nonec replied:
I just tried and it seems to work without the timeout.

andrewspinks commented on Tests/Swift/e2e/PactConsumerSwiftTests/PactSpecs.swift:
```diff
@@ -299,7 +299,7 @@ class PactSwiftSpec: QuickSpec {
             }
-            
+
             if(testSetup == TestSetup.Ruby) {
```
If the eachLike matcher is working properly, then the if statement can be removed 
```if(testSetup == TestSetup.Ruby) {```

nonec replied:
I've done that

andrewspinks commented on PactConsumerSwift.podspec:
```diff
@@ -1,7 +1,7 @@
   s.name         = "PactConsumerSwift"
   s.module_name  = "PactConsumerSwift"
-  s.version = "0.5.3"
+  s.version      = "0.5.5"
```
I haven't really thought through versions yet, but this is a huge change, so it would probably be a major version increment and probably released as beta first.

nonec replied:
Sure, it was just to have something updated

andrewspinks commented on PactConsumerSwift.podspec:
```diff
@@ -20,14 +20,17 @@ Pod::Spec.new do |s|
-  s.source       = { :git => "https://github.com/DiUS/pact-consumer-swift.git", :tag => "v#{s.version}" }
-  s.source_files = 'Sources/**/*.swift'
-  s.resources    = 'scripts/start_server.sh', 'scripts/stop_server.sh'
+  s.source       = { :git => "https://github.com/nonec/pact-consumer-swift.git", :tag => "v#{s.version}" }
```
Will need to switch this url back before merging

nonec replied:
Of course

andrewspinks commented on Sources/Matcher.swift:
```diff
@@ -14,19 +14,22 @@ protocol Matchers {
-open class Matcher: NSObject {
+public class Matcher: NSObject {
+  @objc
+  static var matchers: Matchers = RubyMatcher()
```
This is temporary, while we support two implementations, but I wonder if it might still be better to provide a specific method to change it, rather than exposing the variable. e.g.
```
public class func setDefaultMatcher()
```
We can probably also default to the Native ones.

nonec replied:
Yeah, I agree. I'll think of a nicer way and would also prefer rust to be the default.

andrewspinks commented on Sources/MockService.swift:
```diff
@@ -15,18 +14,31 @@ open class MockService: NSObject {
+              matchers: Matchers) {
     self.pact = Pact(provider: provider, consumer: consumer)
     self.mockServer = mockServer
+    Matcher.matchers = matchers
```
Could setting the correct matchers happen in the RubyMockServer / NativeMockServer rather than having an extra parameter here?

nonec replied:
if I remember it correctly I thought about it but didn't want to make this big change. I'll check up on it.

andrewspinks commented on Sources/Native/NativeMockServerWrapper.swift:
```diff
@@ -55,6 +52,19 @@ class NativeMockServerWrapper: MockServer {
+  private func createServerOnUnusedPort(withJson sanitizedString: String) -> Int32 {
+    var result = create_mock_server_ffi(sanitizedString, port)
+    var count = 0
+    while result == -4 && count < 25 {
```
Might be time to pull these hardcoded response codes into constants or a struct?

nonec replied:
Sounds good 👍

andrewspinks commented on Sources/Native/PactBodyBuilder.swift:
```diff
@@ -19,38 +19,57 @@ class PactBodyBuilder {
     case let dictionary as JSONEntry:
-      return processDictionary(dictionary, path: path)
+      result = processDictionary(dictionary, path: path)
+    case let matcher as MinTypeMatcher:
```
Can you add some tests to PactQueryBuilderSpec to cover these scenarios.

nonec replied:
I added a test for this to PactBodyBuilderSpec (since it was PactBodyBuilder)

andrewspinks commented on Sources/Native/PactBodyBuilder.swift:
```diff
@@ -19,38 +19,57 @@ class PactBodyBuilder {
+    return result
+  }
+
+  private func eachLikeMatchingRule(path: String, element: Any) -> PathWithMatchingRule? {
```
Is this used?

nonec replied:
good call. it doesn't seem like it. ~I'll remove it.~

edit: did remove it 👍

andrewspinks commented on Sources/Native/PactBodyBuilder.swift:
```diff
@@ -19,38 +19,57 @@ class PactBodyBuilder {
       let processedSubElement = processElement(path: "\(path)[\(index)]", element: arrayValue)
-      processedArray.append(processedSubElement.0)
-      matches = matches.merge(dictionary: processedSubElement.1)
+      if let eachLikeElement = arrayValue as? MinTypeMatcher {
```
Sorry, I don't full understand this. What is `numberOfBodyElements` for? Can you add some tests around this please.

nonec replied:
I redid this part.

andrewspinks commented on .swift-version:
```diff
@@ -1 +0,0 @@
-4.0
```
Can you add this back in and up to 5.0

nonec replied:
This will be done, once I merged the changes from current master

andrewspinks commented on Sources/Pact.swift:
```diff
@@ -14,7 +14,11 @@ public class Pact {
 
-  func payload() -> [String: Any] {
+  public func addInteractions(_ interaction: Interaction) {
+    self.interactions.append(interaction)
```
What was the reason for exposing this?

nonec replied:
Long story short: since I need it for my test setup. 
Thinking about it I'm sure I can find a solution without needing it here

andrewspinks reviewed (changes requested):
@nonec Thanks so much for this PR. I haven't looked into this for a while, so its been slow for me to remember what was going on in this branch. 

I have a few comments below, and also had a couple of questions:
- Was the only reason to split the rust binary because of the file size? I used git lfs last time https://git-lfs.github.com/ While I don't think there is any problem in theory in splitting it as you have, it might just be more tricky to upgrade the libs in the future.
- How did you end up compiling the libs (for when we wan to do it again. I originally used cargo lipo)
- Have you had any issues with bitcode settings? When I was doing this ages ago, rust didn't support bitcode and I didn't want to enforce disabling bitcode on users. I'm not sure if that is still the case.

Also, could you squash some of your commits? (but definitely leave the merge commit from master as its own)

nonec commented:
Hey Andrew, thanks a lot for the feedback. It’ll take me a while to answer since I’m afk for a few weeks 😉 I’ll get back to you / fix and explain stuff afterwards.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

macblazer committed:
Completely removed Alamofire in favor of simple networking calls. (#51)

* Completely removed Alamofire in favor of simple networking calls.

* Simplified the AnimalServiceClient and the PactVerificationService.

* Moved session property near networking functions.

surpher committed:
Refactor pact mock server network calling code and upgrade to swift 5 (#53)

* Gets the project built

Update base proejct to Swift5 and targets to use undefined swift version (inherit from base project)

* Update travisCI config to run Xcode10.2 with swift 5

* Fixes the enum empty value error

* Update README [skip-ci]

Install pact-mock-service using homebrew

* Sepcify Swift version 5 for tvOS and macOS

* Remove unused import

* Update Swiftlint rules

- set `line_length` warning and error
- exclude keywords for HTTP Methods from identifier_name

* Add Separated NetworkLayer

Removes dependency on Alamofire

* Removes Alamofire

* Remove Alamofire from Xcode Project and dep managers

* Add Network Layer files to list of Compiled Sources

* Clean up Project file

* Update CI step to run mock from brew tap

* Update Swiftlint rules

- set `line_length` warning and error
- exclude keywords for HTTP Methods from identifier_name

* Add Separated NetworkLayer

Removes dependency on Alamofire

* Removes Alamofire

* Remove Alamofire from Xcode Project and dep managers

* Add Network Layer files to list of Compiled Sources

* Clean up Project file

* Update CI step to run mock from brew tap

* Add flag to enable networkLogging

Initialising `PactVerificationService` wtih `networkLogging = true` will display calls and responses to the Pact Mock Service to help with debugging of url requests and data passed between the client and provider.

* Update build script to exit at non-zero status from a cmd

* Update CI to use pact-foundation tap for pact-ruby-standalone

* Resolves outdated SwiftPM packages

* Documentation style comments for public methods

* Update README.md

* Remove redundant hardcoded header value

* Set code indentation to 2 spaces

* Changed indentation to 2 spaces

* Remove trailing empty lines

* Remove duplicated swiftlint identifer_name rules

* Rename Endpoint API to PactMockServiceAPI

* Set urlRequest header values regardless

* Remove redundant case for setting nullable parameters

* Update documentation comments

* Rename `additionalHeaders` to `headers`

* Changed return type in demo AnimalServiceClientApi

Returning an error object on failed URLResponse. When response data is missing, return an error object instead of 500.

* Using Swift5 Response<String, Error>

andrewspinks committed:
Remove quick dependency from requirements. It is only used for internal tests. (#58)

macblazer committed:
Fixes bitcode related issue when used from CocoaPods.

surpher committed:
Merge pull request #60 from macblazer/feature/fix-bitcode

Fixes bitcode related issue when used from CocoaPods.

andrewspinks committed:
Cleanup build settings so ci build can be more easily reproduced locally. (#59)

Commit:
update changelog for release 0.6.0

Commit:
bump cocoapods version to 0.6.0

surpher committed:
Removes watchOS target and scheme (#64)

* Add script to check fro dependencies

* Remove watchOS target and scheme

* Remove unnecessary pre- and post- script commands

* Moves linting step before compiling

* Remove swiftlint —strict argument

* Display Error in Issue Navigator when mock service not installed

Fails the test build and displays error in Issue Navigator when pact-mock-service is not installed.

surpher committed:
Remove unused .ruby-version file (#65)

andrewspinks committed:
Cleanup access controls on classes to hide internal concepts.

surpher committed:
Merge pull request #67 from DiUS/cleanup-access-controls

Cleanup access controls on classes to hide internal concepts.

surpher committed:
chore(lint): moved linting into a portable script (#66)

* Move swiftlint build phase into a bash script

* Improve script that checks for Test dependencies

* Remove swiftlint from build script

* Display help link with swiftlint wraning

surpher committed:
fix(lint): Parameter alignment

Fix the vertical parameter alignment swiftlint violation

andrewspinks committed:
test: Add test around pact mock service to make refactors easier.

nonec committed:
readded better handling (and controlling) for pact writing

andrewspinks committed:
fix: Revert "Refactor pact mock server network calling code and upgrade to swift 5 (#53)" which introduced a bug where messages from the mock server were no longer being returned.

This reverts commit 8688fdc739c95d3f981ba4a63e3ca601ae873947.

# Conflicts:
#	.travis.yml
#	Cartfile.resolved
#	Package.resolved
#	Package.swift
#	PactConsumerSwift.xcodeproj/project.pbxproj
#	Sources/Interaction.swift
#	Sources/JSONParameterEncoder.swift
#	Sources/MockService.swift
#	Sources/PactVerificationService.swift
#	Sources/Router.swift

andrewspinks committed:
chore: Fix test failures not failing travis build.

andrewspinks committed:
chore: Only run coverage report when tests pass

marturi committed:
fix(objc): add escaping modifier to allow for testComplete function to be called… (#71)

* Add escaping modifier to allow for testComplete function to be called in an async block

* Update MockService.swift

* Addressed PR feedback

andrewspinks committed:
chore: Bumping version to 0.6.1

andrewspinks committed:
chore: convert release script to be a bash script #46

surpher committed:
fix: Bump dependencies to support swift 5. Also removes obsolete references to Result.framework.

* fix: Remove unused import

* (fix): Convert non-mutable vars to `let`

* (fix): Remove unused framework from copying to bundle

* (chore): Cleanup targets’ build settings and build phases

Remove references to redunant `Result.framework` that has been removed from the project a few minor versions ago.

* (chore): Bump up the simulator OS version in CI

* (chore): Bump up the dependencies version

Bumps up `BrightFutures.framework` version to `8.0` to support Swift 5.

* (chore) Bump up iOS simulator version

* (fix) Revert Cocoapods version number

The version number for Cocapods is bumped up using a script. It should have not been done manually.

andrewspinks committed:
chore: Bumping version to 0.6.2

andrewspinks committed:
Revert "chore: Bumping version to 0.6.2" due to cocoapods issues.

This reverts commit 58775a223f442fe1ae9ec414991583e7c6db3e00.

andrewspinks committed:
chore: Change minimum swift version to 4.2 due Nimble requirements. Also fix cocoapods configuration.

nonec committed:
increased tries for a unique server port generation

nonec committed:
added a sleep if the server number is in use

nonec committed:
remember (and avoid) already used ports

nonec committed:
reformatting

nonec committed:
checking the port before opening the server

nonec committed:
increased random range and added comment

nonec committed:
stronger typing for eachlike

alannichols committed:
style: add convenience init method to MockService which defaults error reporting but allows configuration of the verifier. (#77)

andrewspinks committed:
chore: Bumping version to 0.7.1

TomasLinhart committed:
feat: add support for macCatalyst (#78)

* Add support for macCatalyst
Removes all referenced frameworks because having them in FRAMEWORK_SEARCH_PATHS is enough. Introduce a new variable CARTHAGE_SUBFOLDER that is different in case of macOS, so a correct framework can be used in case of macCatalyst.

* Copy correct frameworks

georgescumihai committed:
feat: allow self signed ssl certificates (#74)

* Allow insecure certifiactes in Pact verification system

* Fix documentation

* reafactor: Passing `allowInsecureCertificates` as parameter at init

* Fix warnings

Co-authored-by: Marko Justinek <marko.justinek@gmail.com>
Co-authored-by: piellarda <piellarda@gmail.com>

nonec committed:
Merge remote-tracking branch 'upstream/master'

nonec committed:
Merge branch 'feature/nativeWrapperUpdate' into feature/nativeWrapperUpdatePR

nonec committed:
Merge branch 'master' into feature/nativeWrapperCurrent

# Conflicts:
#	.travis.yml
#	CONTRIBUTING.md
#	Cartfile
#	Cartfile.private
#	Cartfile.resolved
#	Gemfile
#	Gemfile.lock
#	Package.resolved
#	Package.swift
#	PactConsumerSwift.podspec
#	PactConsumerSwift.xcodeproj/project.pbxproj
#	Sources/Interaction.swift
#	Sources/Matcher.swift
#	Sources/MockService.swift
#	Sources/Ruby/RubyMockServer.swift
#	Tests/PactConsumerSwiftTests/PactSpecs.swift
#	Tests/Swift/Unit/RubyInteractionAdapterSpec.swift
#	scripts/build.sh

nonec committed:
adjusted for some tests

nonec committed:
readded swiftjson for tests and removed watch target

nonec committed:
removed unnecessary casts

nonec committed:
readded swiftjson for mismatches parsing

nonec committed:
rearranged matcher code for reuse

nonec committed:
readding convenience init with matcher

nonec committed:
convenience

nonec committed:
set generate to Any to work with numbers

nonec commented:
> @nonec Thanks so much for this PR. I haven't looked into this for a while, so its been slow for me to remember what was going on in this branch.
> 
> I have a few comments below, and also had a couple of questions:
> 
> * Was the only reason to split the rust binary because of the file size? I used git lfs last time https://git-lfs.github.com/ While I don't think there is any problem in theory in splitting it as you have, it might just be more tricky to upgrade the libs in the future.
> * How did you end up compiling the libs (for when we wan to do it again. I originally used cargo lipo)
> * Have you had any issues with bitcode settings? When I was doing this ages ago, rust didn't support bitcode and I didn't want to enforce disabling bitcode on users. I'm not sure if that is still the case.
> 
> Also, could you squash some of your commits? (but definitely leave the merge commit from master as its own)

Hey @andrewspinks , 
finally found some time to work on this again. I'll update this pull request (mainly to the changes from master) but I'll first go through your questions.
- git-lfs sounds like a good idea and I'll look into this for the pull request. while my solution works it's more work and gives more opportunity for error. 
- I looked into the reference project for the rust mock server and used the commands from the builds scripts. I'm not a 100% sure anymore but I also this I did it with cargo limo
- I can't remember that I had problems with bitcode

nonec committed:
reordering test folder

nonec committed:
renaming ruby mock service

nonec committed:
removed unused method

nonec committed:
removed unrelated method

was needed for outside requirement and should be implemented there

nonec committed:
further adoptions of testing paths

nonec committed:
adaptation of MatchingRule for ruby

nonec committed:
updated minType handling

nonec committed:
enabled test for rust

nonec committed:
increase count to verify multiple elements

nonec committed:
removed timeout

nonec committed:
added specific matchers

nonec committed:
removed unused timeout

nonec committed:
fixed a few swift build errors

nonec committed:
removed debug prints

nonec committed:
updated podspec

nonec committed:
project file cleanup

nonec committed:
added swift include paths

Commit:
using swift package for swift build

Commit:
cleaned up build phases

Commit:
updated build versions

Commit:
removed non existing source

Commit:
corrected to existing version

Commit:
added bootstrap command

Commit:
Revert "added bootstrap command"

This reverts commit 804de2da91d4fadb8286c421b097e2059beff6db.

Commit:
setting to existing simulator

Commit:
added some additional debug for failure to travis

Commit:
moved dependencies and bootstrapping to travis

Commit:
added carthage cache

Commit:
reduced log output size

nonec commented:
> * Was the only reason to split the rust binary because of the file size? I used git lfs last time https://git-lfs.github.com/ While I don't think there is any problem in theory in splitting it as you have, it might just be more tricky to upgrade the libs in the future.

I just realised I maybe even tried git-lfs before, but it's not possible to push with git-lfs to a fork.

Commit:
reduced debug further

Commit:
moved building to travis config file

Commit:
updated phone

Commit:
only write matching rules once

Commit:
simple test for pact body builder

Commit:
minor changes on pact file structure

Commit:
Merge branch 'feature/nativeWrapperCurrent' into feature/nativeWrapperPR

surpher commented on Brewfile:
```diff
@@ -0,0 +1,4 @@
+tap 'pact-foundation/pact-ruby-standalone'
```
To avoid having these changes that were not made by you in the PR I would encourage you to rebase `origin:master` onto your `fork:branchPR`, then force push (onto your `fork:branchPR`!!!) the new history. 

That way any of the changes made to origin will not show up in your PR.

nonec replied:
TODO: (for me)
yes, I agree. I'll have to see when I have time for this. Since I think this is what I did the last time.

surpher commented on CHANGELOG.md:
```diff
@@ -1,5 +1,74 @@
+# 0.7.1 - Bugfix Release
```
Same applies here regarding rebasing on your PR branch. You haven't made this changes, have you? Therefore they should not be part of your PR. 🤷

nonec replied:
Agreed, like above

surpher commented on Cartfile:
```diff
@@ -1,4 +1,3 @@
-github "SwiftyJSON/SwiftyJSON" ~> 4.0
+github "Thomvis/BrightFutures" ~> 8.0
+github "Quick/Nimble" ~> 8.0
+github "SwiftyJSON/SwiftyJSON" ~> 5.0.0
```
After having `Codable` available to us since Swift 4, I'd encourage you to avoid bringing in more dependencies for things that Swift can do for by just using what `Foundation` provides.

nonec replied:
Also something I agree on. Original the branch was done by @andrewspinks . When I first started working on this my goal was to get it to a point that it can be merged back. 
Since Codable came up I thought about this, but the top priority was to get everything else working right first.

surpher commented on Cartfile.private:
```diff
@@ -1 +1,2 @@
-github "Quick/Quick" ~> 1.1
+github "Quick/Quick" ~> 2.0
+github "AliSoftware/OHHTTPStubs"
```
I know this is not you @nonec but if you run `carthage update` now it will pull version 9.0.0 and it will break current implementation of tests as this dependency dropped `OH` prefix from all its class names. All it will bring is a headache. :|

nonec replied:
That is a headache, yes :|

surpher commented on Sources/Native/DictionaryExtensions.swift:
```diff
@@ -3,8 +3,8 @@ import Foundation
     var newDictionary: [Key: Value] = self
-    for (key, value) in dictionary {
-      newDictionary[key] = value
+    for (dictKey, value) in dictionary {
```
Nitpicking here... The `newDictionary` already defines it as `[Key: Value]`, following the convention would make sense.

nonec replied:
Are you talking about the type `[Key: Value]` or the naming `(dictKey, value)`?

surpher commented on Sources/Native/NativeMockServerWrapper.swift:
```diff
@@ -2,102 +2,206 @@ import Foundation
+    import SwiftPactServer
+#endif
+
+public class NativeMockServerWrapper: MockServer {
```
It is really hard to read through this class where public and private declarations are all over the place. It is hard to read and learn about interface the caller needs to interact with.

nonec replied:
I think it's mostly private except the `MockServer` protocol implementations.
What would you do to improve it?

surpher commented on Sources/Native/NativeMockServerWrapper.swift:
```diff
@@ -2,102 +2,206 @@ import Foundation
+
+public class NativeMockServerWrapper: MockServer {
+    private lazy var port: Int32 = {
+        return findUnusedPort()
```
If this `findUnusedPort()` is only called here, why not have the logic encapsulated here? it's a private var too.

nonec replied:
TODO:
yes. I think this grew and probably was used somewhere else before.

surpher commented on Sources/Native/NativeMockServerWrapper.swift:
```diff
@@ -2,102 +2,206 @@ import Foundation
+    private let pactDir: String = ProcessInfo.processInfo.environment["pact_dir"] ?? "/tmp/pacts"
+    private let shouldWritePacts: Bool
+
+    public init(shouldWritePacts: Bool = false) {
```
is `shouldWritePacts` used to have a dry run on tests when it's set to `false`? Or does it have something to do with Filesystem?

nonec replied:
If I remember it correctly this was needed for the setup I'm using.
I think part of the problem was this is failing on devices and it only worked for simulators. So I had to control a bit when to write for being able to automate this in our building pipeline.

surpher commented on Sources/Native/NativeMockServerWrapper.swift:
```diff
@@ -2,102 +2,206 @@ import Foundation
+
+                // iOS json generation adds extra backslashes to "application/json" --> "application\\/json"
+                // causing the MockServer to fail to parse the file.
+                let sanitizedString = jsonString!.replacingOccurrences(of: "\\/", with: "/")
```
This sanitisation is occurring in a few places, could pull it out into a reusable method or a String extension.

nonec replied:
TODO: 
always a good idea

surpher commented on Sources/Ruby/RubyMockServer.swift:
```diff
@@ -1,33 +1,33 @@
     case write([String: [String: String]])
 
-    var method: HTTPMethod {
+    var method: String {
```
HTTPMethod.rawValue should be the only place where the mapping of these should occur.

nonec replied:
I think HTTPMethod was from Alamofire which I removed as a dependency here.
So what would be the todo here?

surpher commented on Sources/Ruby/RubyMockServer.swift:
```diff
@@ -46,25 +46,38 @@ open class RubyMockServer: MockServer {
     // MARK: URLRequestConvertible
     func asURLRequest() throws -> URLRequest {
-      let url = try Router.baseURLString.asURL()
+      guard let url = URL(string: Router.baseURLString) else { throw NSError(domain: "", code: 1, userInfo: nil) }
```
A new merge on the origin:master made Router method to `throw`. This is kind of out of date now. :|

nonec replied:
TODO:
Yeah, I'll have to see about that

surpher commented on Package.swift:
```diff
@@ -1,30 +1,32 @@
+    .package(url: "https://github.com/Quick/Quick.git", from: "2.1.0"),
+    .package(url: "https://github.com/SwiftyJSON/SwiftyJSON.git", from: "5.0.0"),
+    .package(url: "https://github.com/Thomvis/BrightFutures.git", from: "8.0.0"),
+    .package(url: "https://github.com/nonec/SwiftPactServer.git", from: "1.0.0"),
```
Instead of having yet another git repo you rely on due to SPM, you could have it as part of the current project and define dependency as `.package(path: "path/to/folder/containing/SwiftPactServer")`

nonec replied:
I remember thinking about that. I think part of the problem was having big files in that and I couldn't use git-lfs on a fork because that would count into the space of the origin.
I'm not a 100% sure, but I think this would be something to resolve if we ever get this to a mergeable state.

surpher commented:
Great job in doing this @nonec ! 👍 I see it still supports pact spect v2? Have you looked into implementing v3 to match the [rust pact_mock_server](https://github.com/pact-foundation/pact-reference/tree/master/rust)?

I had a scroll through but there is so much of it and there are so many changed files I know are not files you changed. It is very very hard to provide feedback because of that. I did leave a few comments that I _think_ you made and I hope you can take it onboard (or take with you and apply on other projects). 

If you do manage to merge/rebase and align your PR with current `DiUS/pact-consumer-swift:master` so we can see only your changes, I'm happy to provide more constructive and in-depth feedback.

surpher commented on Sources/Native/NativeMockServerWrapper.swift:
```diff
@@ -2,102 +2,206 @@ import Foundation
+        }
+    }
+
+    public func verify(_ pact: Pact) -> Future<String, PactError> {
```
You're passing `Pact` into `verify()` yet it is not used in the method.

nonec replied:
That is true. It's part of the protocol (and the ruby implementation uses it). I could just ignore it here (or move away from the ruby alternative and change the protocol)
TODO:

surpher commented on Sources/Native/NativeMockServerWrapper.swift:
```diff
@@ -2,102 +2,206 @@ import Foundation
+
+    private func writeFile() -> Int32 {
+        guard shouldWritePacts, checkForPath() else {
+            return 4
```
It's really hard to understand what this error would be. I'd suggest you use an enum for these. Something like:

```
enum VerifyResult {
  case failToWrite
  case failToConnect
  case successWrite
  case successVerifyOnly
  case unknown(String?)
}
```

Naming could be much better for this example, but you get an idea. 
That way when you switch through the case you can read it and understand what is happening (case `.successVerifyOnly` rather than dealing with `4`.

nonec replied:
TODO:
agreed. same is valid for the error handling in `setup(_ pact: Pact)`. I think there was a comment on that before as well.

nonec reviewed (commented):
> Great job in doing this @nonec ! 👍 I see it still supports pact spect v2? Have you looked into implementing v3 to match the [rust pact_mock_server](https://github.com/pact-foundation/pact-reference/tree/master/rust)?
> 
> I had a scroll through but there is so much of it and there are so many changed files I know are not files you changed. It is very very hard to provide feedback because of that. I did leave a few comments that I _think_ you made and I hope you can take it onboard (or take with you and apply on other projects).
> 
> If you do manage to merge/rebase and align your PR with current `DiUS/pact-consumer-swift:master` so we can see only your changes, I'm happy to provide more constructive and in-depth feedback.

Thanks for taking the time looking at this. I'm sorry this got outdated, again.
At the moment the biggest work to get this started was done by @andrewspinks . I only picked this up (twice), updated it and get it to run. 
So yes, atm it supports v2, yes. I'd really like for it to support v3.

Tbh I'm not sure if I'll do this dance once more or I'll just pick up my learnings and start over with something that only uses the native wrapper and implements v3.
Not sure when I'll have time to get into this, either way :|

So for now I tried to answer your comments and added TODOs for myself to go through, when there's time on my side.

andrewspinks commented:
Most of the learnings from this PR / branch have been taken and used in a new project here: https://github.com/surpher/PactSwift The idea is to eventually move that to the pact-foundation once it has stabilized. @nonec it would be awesome if you could take a look at that new project when if you get some spare time. I'm going to close this PR to save it getting even more out of date.

nonec commented:
> Most of the learnings from this PR / branch have been taken and used in a new project here: https://github.com/surpher/PactSwift The idea is to eventually move that to the pact-foundation once it has stabilized. @nonec it would be awesome if you could take a look at that new project when if you get some spare time. I'm going to close this PR to save it getting even more out of date.

Oh, that looks great! I hope I'll have some time to look into it soon. Thanks for the heads up!
